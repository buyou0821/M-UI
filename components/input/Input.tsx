import React, { forwardRef, useRef, useState, useLayoutEffect } from 'react';
import clsx from 'clsx';
import { usePrefixCls } from '../_util/hooks';
import omit from '../_util/omit';
import { Omit } from '../_util/type';

export type InputType = 'text' | 'number' | 'password' | 'textarea';

interface InputPorps extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'prefix'> {
  className?: string;
  errorMessage?: string;
  label?: string;
  labelFloat?: boolean;
  icon?: React.ReactNode;
  prefix?: React.ReactNode;
  suffix?: React.ReactNode;
  type?: InputType;
}

const Input = forwardRef((props: InputPorps, ref: React.RefObject<HTMLDivElement>) => {
  const {
    disabled,
    errorMessage,
    label,
    labelFloat,
    prefix,
    suffix,
    type = 'input',
    icon,
    className,
  } = props;
  const inputRef = useRef<HTMLInputElement>(null);
  const [focused, setFocused] = useState<boolean>(false);
  const [isFloat, setIsFloat] = useState<boolean>(false);

  const handleFocus = () => {
    setFocused(true);
  };

  const handleBlur = () => {
    setFocused(false);
  };

  const prefixCls = usePrefixCls('input');
  const inputClasses = clsx(prefixCls);
  const wrapperClasses = clsx(
    `${prefixCls}__wrapper`,
    {
      [`${prefixCls}--error`]: errorMessage !== undefined,
      [`${prefixCls}--disabled`]: disabled,
    },
    className,
  );
  const iconClasses = clsx(`${prefixCls}__icon`, {
    [`${prefixCls}__icon--focused`]: focused,
  });
  const contentClasses = clsx(`${prefixCls}__content`, {
    [`${prefixCls}--focus`]: focused,
  });
  const mainClasses = clsx(`${prefixCls}__main`, {});
  const labelClasses = clsx(`${prefixCls}__label`, {
    [`${prefixCls}__label--focused`]: isFloat,
  });

  useLayoutEffect(() => {
    if (focused || (inputRef.current && inputRef.current.value.length > 0) || !labelFloat) {
      setIsFloat(true);
    }
    return () => {
      setIsFloat(false);
    };
  }, [focused]);

  const Label = label ? <label className={labelClasses}>{label}</label> : null;
  const Icon = icon ? <div className={iconClasses}>{icon}</div> : null;
  const Prefix = prefix ? <span className={`${prefixCls}__prefix`}>{prefix}</span> : null;
  const Suffix = suffix ? <span className={`${prefixCls}__suffix`}>{suffix}</span> : null;

  const inputProps: any = omit(props, [
    'className',
    'prefix',
    'suffix',
    'errorMessage',
    'label',
    'labelFloat',
    'placeholder',
    'icon',
  ]);

  if (props.placeholder && isFloat) {
    inputProps.placeholder = props.placeholder;
  } else {
    delete inputProps.placeholder;
  }

  return (
    <div ref={ref} className={wrapperClasses}>
      {Icon}
      <div className={mainClasses}>
        <div className={contentClasses}>
          {Label}
          {isFloat && Prefix}
          <input
            ref={inputRef}
            className={inputClasses}
            type={type}
            onFocus={handleFocus}
            onBlur={handleBlur}
            {...inputProps}
          />
          {isFloat && Suffix}
        </div>
        {errorMessage && (
          <div className={`${prefixCls}__info`}>
            <div className={`${prefixCls}__errorMessage`}>{errorMessage}</div>
          </div>
        )}
      </div>
    </div>
  );
});

export default Input;
