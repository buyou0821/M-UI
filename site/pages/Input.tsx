import React from 'react';
import { Input, Icon } from 'components';

export default () => {
  return (
    <div style={{ padding: 50 }}>
      <Input />
      <Input errorMessage="请输入密码" />
      <Input label="prefix" prefix="$" />
      <Input label="labelFloat" labelFloat />
      <Input label="labelFloat prefix" labelFloat prefix="KG" />
      <Input label="labelFloat prefix suffix" labelFloat prefix="KG" suffix="元" />
      <Input
        type="password"
        placeholder="请输入姓名"
        label="labelFloat prefix suffix password"
        labelFloat
        prefix="KG"
        suffix="元"
      />
      <Input label="placeholder" placeholder="请输入姓名" />
      <Input
        label="iconInput"
        icon={<Icon material="person" />}
        labelFloat
        placeholder="请输入姓名"
      />
      <Input label="iconInput" icon={<Icon material="person" />} placeholder="请输入姓名" />
      <Input disabled label="disabled" icon={<Icon material="person" />} placeholder="请输入姓名" />
      <Input disabled errorMessage="请输入密码" label="disabled" icon={<Icon material="person" />} placeholder="请输入姓名" />
      <Input labelFloat errorMessage="请输入密码" label="disabled" icon={<Icon material="person" />} placeholder="请输入姓名" />
    </div>
  );
};
