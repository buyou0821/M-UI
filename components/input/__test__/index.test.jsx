import React from 'react';
import { mount } from 'enzyme';
import Input from '..';

describe('Input', () => {
  it('should support maxLength', () => {
    const wrapper = mount(<Input maxLength={10} placeholder="input..." />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should support prefix and suffix', () => {
    const wrapper = mount(<Input label="Regular" prefix="$" suffix="å…ƒ" />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Textarea', () => {
  it('should support disabled', () => {
    const wrapper = mount(<Input type="textarea" disabled />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should support maxLength', () => {
    const wrapper = mount(<Input type="textarea" maxLength={10} />);
    expect(wrapper).toMatchSnapshot();
  });
});
