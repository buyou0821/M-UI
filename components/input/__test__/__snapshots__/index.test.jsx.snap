// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should support maxLength 1`] = `
<ForwardRef
  maxLength={10}
  placeholder="input..."
>
  <div
    className="ty-input__wrapper"
  >
    <div
      className="ty-input__main"
    >
      <div
        className="ty-input__content"
      >
        <input
          className="ty-input"
          maxLength={10}
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="input..."
          type="text"
        />
      </div>
      <div
        className="ty-input__info"
      >
        <div
          className="ty-input__errorMessage"
        />
      </div>
    </div>
  </div>
</ForwardRef>
`;

exports[`Input should support prefix and suffix 1`] = `
<ForwardRef
  label="Regular"
  prefix="$"
  suffix="元"
>
  <div
    className="ty-input__wrapper"
  >
    <div
      className="ty-input__main"
    >
      <div
        className="ty-input__content"
      >
        <label
          className="ty-input__label ty-input__label--focused"
        >
          Regular
        </label>
        <span
          className="ty-input__prefix"
        >
          $
        </span>
        <input
          className="ty-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          type="text"
        />
        <span
          className="ty-input__suffix"
        >
          元
        </span>
      </div>
    </div>
  </div>
</ForwardRef>
`;

exports[`Textarea should support disabled 1`] = `
<ForwardRef
  disabled={true}
  type="textarea"
>
  <div
    className="ty-input__wrapper ty-input--disabled"
  >
    <div
      className="ty-input__main"
    >
      <div
        className="ty-input__content"
      >
        <ForwardRef
          className="ty-input ty-input__textarea"
          disabled={true}
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          type="textarea"
        >
          <textarea
            className="ty-input ty-input__textarea"
            disabled={true}
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            type="textarea"
          />
        </ForwardRef>
      </div>
    </div>
  </div>
</ForwardRef>
`;

exports[`Textarea should support maxLength 1`] = `
<ForwardRef
  maxLength={10}
  type="textarea"
>
  <div
    className="ty-input__wrapper"
  >
    <div
      className="ty-input__main"
    >
      <div
        className="ty-input__content"
      >
        <ForwardRef
          className="ty-input ty-input__textarea"
          maxLength={10}
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          type="textarea"
        >
          <textarea
            className="ty-input ty-input__textarea"
            maxLength={10}
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            type="textarea"
          />
        </ForwardRef>
      </div>
      <div
        className="ty-input__info"
      >
        <div
          className="ty-input__errorMessage"
        />
      </div>
    </div>
  </div>
</ForwardRef>
`;
